map.go

```diff
@@ -58,7 +58,7 @@ func deepMap(dst, src reflect.Value, visited map[uintptr]*visit, depth int, conf
 			}
 			fieldName := field.Name
 			fieldName = changeInitialCase(fieldName, unicode.ToLower)
-			if v, ok := dstMap[fieldName]; !ok || (isEmptyValue(reflect.ValueOf(v), !config.ShouldNotDereference) || overwrite) {
+			if _, ok := dstMap[fieldName]; !ok || (!isEmptyValue(reflect.ValueOf(src.Field(i).Interface()), !config.ShouldNotDereference) && overwrite) || config.overwriteWithEmptyValue {
```

Comment:
Unfortunately, this change would break our use case, consider the following scenario 
```go
	dst := map[string]interface{}{
		"foo":   nil,
		"bar":   1,
	}
	src := struct {
		Foo   *string
		Bar   int
	}{
		Foo: &"hello",
		Bar: 42,
	}
	if err := mergo.Map(&dst, src); err != nil {
		t.Error(err)
	}
```
we expect  *dst["foo"] == "hello" and  dst["bar"] == 1

Reply:
@frankdvd Your use case seems to be an unintended consequence. Mergo shouldn't flatten structures while mapping.